tests/test_browser.py

```diff
@@ -0,0 +1,77 @@
+from os import getenv
+from pathlib import Path
+from shutil import which
+from subprocess import check_output
+from uuid import uuid4
+
+from PIL import Image, ImageChops
+from playwright.sync_api import expect
+
+HERE = Path(__file__).absolute().parent
+
+CONTAINER_ID = getenv("CONTAINER_ID", "test")
+JUPYTER_HOST = getenv("JUPYTER_HOST", "http://localhost:8888")
+JUPYTER_TOKEN = getenv("JUPYTER_TOKEN", "secret")
+
+
+def compare_screenshot(test_image, threshold=2):
+    # Compare images by calculating the mean absolute difference
+    # Images must be the same size
+    # threshold: Average difference per pixel, this depends on the image type
+    # e.g. for 24 bit images (8 bit RGB pixels) threshold=1 means a maximum
+    # difference of 1 bit per pixel per channel
+    reference = Image.open(HERE / "reference" / "desktop.png")
+    test = Image.open(test_image)
+
+    # Absolute difference
+    # Convert to RGB, alpha channel breaks ImageChops
+    diff = ImageChops.difference(reference.convert("RGB"), test.convert("RGB"))
+    diff_data = diff.getdata()
+
+    m = sum(sum(px) for px in diff_data) / diff_data.size[0] / diff_data.size[1]
+    assert m < threshold
+
+
+# To debug this set environment variable HEADLESS=0
+def test_desktop(browser):
+    page = browser
+    page.goto(f"{JUPYTER_HOST}/lab?token={JUPYTER_TOKEN}")
+    page.wait_for_url(f"{JUPYTER_HOST}/lab")
+
+    # JupyterLab extension icon
+    expect(page.get_by_text("Desktop [↗]")).to_be_visible()
+    with page.expect_popup() as page1_info:
+        page.get_by_text("Desktop [↗]").click()
+    page1 = page1_info.value
+    page1.wait_for_url(f"{JUPYTER_HOST}/desktop/")
+
+    expect(page1.get_by_text("Status: Connected")).to_be_visible()
+    expect(page1.locator("canvas")).to_be_visible()
+
+    # Screenshot the desktop element only
+    # May take a few seconds to load
+    page1.wait_for_timeout(5000)
+    # Use a non temporary folder so we can check it manually if necessary
+    screenshot = Path("screenshots") / "desktop.png"
+    page1.locator("canvas").screenshot(path=screenshot)
```

Comment:
This just screenshots the canvas element which contains only the desktop, and not the surrounding NoVNC elements to help minimise changes. We can change this to `page1.screenshot(path=screenshot)` to screenshot the whole page though.